akka-http/src/main/scala/akka/http/javadsl/server/directives/RouteDirectives.scala

```diff
@@ -46,6 +60,83 @@ abstract class RouteDirectives extends RespondWithDirectives {
   }
 
   /**
+   * Composes two 0-argument directives which share the same inner Route. This is equivalent to
+   * first(inner).orElse(second(inner))
+   * @param first the first 0-argument directive
+   * @param second the second 0-argument directive
+   * @param inner the inner route Producer
+   * @return the resulting route
+   */
+  def anyOf(first: JFunction[Supplier[Route], Route], second: JFunction[Supplier[Route], Route], inner: Supplier[Route]): Route = {
+    first.apply(inner).orElse(second.apply(inner))
+  }
+
+  /**
+   * Composes two 1-argument directives (directives whose inner route takes a parameter) which share the same inner route.
+   * This is the same as first(inner).orElse(second(inner)).
+   * This can be used also for directives that take an additional parameter if the directive is partially applied, ie. parameter
+   * @param first the first 1-argument directive
+   * @param second the second 1-argument directive
+   * @param inner the inner route that takes one argument
+   * @tparam A the type of the parameter the directives extract and the inner route takes
+   * @return the resulting route
+   */
+  def anyOf[A](first: JFunction[JFunction[A, Route], Route], second: JFunction[JFunction[A, Route], Route], inner: JFunction[A, Route]): Route = {
```

Comment:
The arguments can take type bound `JFunction[_ >: A]` for both, but double-confirm that those bounds make it into Java; otherwise, we'll have to write these in `.java` files.